functions/src/shared/s3-answers.ts

```diff
@@ -1,51 +1,104 @@
   context_id: { type: 'UTF8', optional: true },
   class_info_url: { type: 'UTF8', optional: true },
   platform_id: { type: 'UTF8', optional: true },
   resource_link_id: { type: 'UTF8', optional: true },
-  type: { type: 'UTF8' },
+  type: { type: 'UTF8', optional: true },
   question_id: { type: 'UTF8' },
   source_key: { type: 'UTF8' },
-  question_type: { type: 'UTF8' },
-  tool_user_id: { type: 'UTF8' },
+  question_type: { type: 'UTF8', optional: true },
+  tool_user_id: { type: 'UTF8', optional: true },
   answer: { type: 'UTF8' },
   resource_url: { type: 'UTF8' },
   remote_endpoint: { type: 'UTF8', optional: true },
-  created: { type: 'UTF8' },
+  created: { type: 'UTF8', optional: true },
   tool_id: { type: 'UTF8' },
-  version: { type: 'UTF8' },
+  version: { type: 'UTF8', optional: true },
 });
 
-export const parquetInfo = (directory: string, answer?: AnswerData | null, _runKey?: string, _resourceUrl?: string) => {
-  let runKey, resourceUrl;
-  if (answer) {
-    runKey = answer.run_key;
-    resourceUrl = answer.resource_url;
+// replaces everything but alphanumeric chars with -
+const escapeUrl = (url: string) => url.replace(/[^a-z0-9]/g, "-")
+
+/**
+ * Returns the filename and directory structure that the parquet file should end up in, given an answer.
+ *
+ * For a logged-in user launching from the portal (LARA or Activity Player):
+ *  answers/[escaped_resource_id]/[platform_id]/[resource_link_id]/[platform_user_id].parquet
+ *
+ * For an anonymous user (LARA or AP):
+ *  answers/[escaped_resource_id]/anonymous/no-resource-link/[run-key].parquet
+ *
+ * @param directory Top-level directory in S3 bucket (partitioned-answers)
+ * @param answer Answer or metadata-only info
+ */
+export const parquetInfo = (directory: string, answer: AnswerMetadata) => {
+  const escaped_resource_id = escapeUrl(answer.resource_url);
+  let filename: string;
+  if (escaped_resource_id && answer.platform_id && answer.resource_link_id && answer.platform_user_id) {
+    filename = `${answer.platform_user_id}.parquet`;
+    return {
+      filename,
+      key: `${directory}/${escaped_resource_id}/${escapeUrl(answer.platform_id)}/${answer.resource_link_id}/${filename}`
+    };
+  } else if (answer.run_key) {
+    filename = `${answer.run_key}.parquet`;
+    return {
+      filename,
+      key: `${directory}/${escaped_resource_id}/anonymous/no-resource-link/${filename}`
+    };
   } else {
-    runKey = _runKey;
-    resourceUrl = _resourceUrl;
+    throw Error(`Cannot create filename`);
   }
-  if (!runKey || !resourceUrl) {
-    throw Error(`Cannot create filename for ${runKey}`);
+}
+
+// returns just the part needed to identify the answer uniquely
+// urls don't need to be escaped because the document name will be hashed
+export const getSyncDocId = (answer: AnswerData) => {
+  if (answer.platform_id && answer.resource_link_id && answer.platform_user_id) {
+    return `${answer.platform_id}_${answer.resource_url}_${answer.platform_user_id}`;
```

Comment:
I'd guess that Firestore will have problems with some characters in its document ids. Since platform_id and resource_url are URLS I'd guess the `:` and `//` and possibly `?` will cause problems.

Reply:
It could maybe be clearer in the code, but this string actually gets hashed before being turned into a firestore document id (in `addSyncDoc` in auto-importer.ts)